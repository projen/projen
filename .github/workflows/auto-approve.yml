# ~~ Generated by projen. To modify, edit .projenrc.ts and run "node ./projen.js".

name: auto-approve
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
jobs:
  approve:
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: contains(github.event.pull_request.labels.*.name, 'auto-approve') && (github.event.pull_request.user.login == 'cdklabs-automation')
    steps:
      - uses: hmarr/auto-approve-action@f0939ea97e9205ef24d872e76833fa908a770363
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
