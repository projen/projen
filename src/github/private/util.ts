export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
