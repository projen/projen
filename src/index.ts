export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
