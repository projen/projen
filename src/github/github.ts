import { GitHubActionsProvider } from "./actions-provider";
import { Dependabot, DependabotOptions } from "./dependabot";
import { GithubCredentials } from "./github-credentials";
import { MergeQueue, MergeQueueOptions } from "./merge-queue";
import { Mergify, MergifyOptions } from "./mergify";
import { PullRequestTemplate } from "./pr-template";
import {
  PullRequestBackport,
  PullRequestBackportOptions,
} from "./pull-request-backport";
import { PullRequestLint, PullRequestLintOptions } from "./pull-request-lint";
import { GithubWorkflow } from "./workflows";
import { Component } from "../component";
import { Project } from "../project";

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;

  /**
   * Options for Mergify.
   *
   * @default - default options
   */
  readonly mergifyOptions?: MergifyOptions;

  /**
   * Whether a merge queue should be used on this repository to merge pull requests.
   * Requires additional configuration of the repositories branch protection rules.
   *
   * @default false
   */
  readonly mergeQueue?: boolean;

  /**
   * Options for MergeQueue.
   *
   * @default - default options
   */
  readonly mergeQueueOptions?: MergeQueueOptions;

  /**
   * Enables GitHub workflows. If this is set to `false`, workflows will not be created.
   *
   * @default true
   */
  readonly workflows?: boolean;

  /**
   * Add a workflow that allows backport of PRs to other branches using labels.
   * When opening a new PR add a backport label to it,
   * and the PR will be backported to the target branches once the PR is merged.
   *
   * Should not be used together with mergify.
   *
   * @default false
   */
  readonly pullRequestBackport?: boolean;

  /**
   * Options for configuring pull request backport.
   *
   * @default - see defaults in `PullRequestBackportOptions`
   */
  readonly pullRequestBackportOptions?: PullRequestBackportOptions;

  /**
   * Add a workflow that performs basic checks for pull requests, like
   * validating that PRs follow Conventional Commits.
   *
   * @default true
   */
  readonly pullRequestLint?: boolean;

  /**
   * Options for configuring a pull request linter.
   *
   * @default - see defaults in `PullRequestLintOptions`
   */
  readonly pullRequestLintOptions?: PullRequestLintOptions;

  /**
   * Choose a method of providing GitHub API access for projen workflows.
   *
   * @default - use a personal access token named PROJEN_GITHUB_TOKEN
   */
  readonly projenCredentials?: GithubCredentials;

  /**
   * The name of a secret which includes a GitHub Personal Access Token to be
   * used by projen workflows. This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   *
   * @default "PROJEN_GITHUB_TOKEN"
   * @deprecated - use `projenCredentials`
   */
  readonly projenTokenSecret?: string;

  /**
   * Download files in LFS in workflows
   *
   * @default true if the associated project has `lfsPatterns`, `false` otherwise
   */
  readonly downloadLfs?: boolean;
}

export class GitHub extends Component {
  /**
   * Returns the `GitHub` component of a project or `undefined` if the project
   * does not have a GitHub component.
   */
  public static of(project: Project): GitHub | undefined {
    const isGitHub = (c: Component): c is GitHub => c instanceof GitHub;
    return project.components.find(isGitHub);
  }

  /**
   * The `Mergify` component configured on this repository
   * This is `undefined` if Mergify is not enabled for this repository.
   */
  public readonly mergify?: Mergify;

  /**
   * The `MergeQueue` component configured on this repository
   * This is `undefined` if merge queues are not enabled for this repository.
   */
  public readonly mergeQueue?: MergeQueue;

  /**
   * Are workflows enabled?
   */
  public readonly workflowsEnabled: boolean;

  /**
   * GitHub API authentication method used by projen workflows.
   */
  public readonly projenCredentials: GithubCredentials;

  /**
   * The GitHub Actions provider used to manage the versions of actions used in steps
   */
  public readonly actions: GitHubActionsProvider;

  private readonly _downloadLfs?: boolean;

  public constructor(project: Project, options: GitHubOptions = {}) {
    super(project);

    this.actions = new GitHubActionsProvider();

    this.workflowsEnabled = options.workflows ?? true;

    this._downloadLfs = options.downloadLfs;

    if (options.projenCredentials && options.projenTokenSecret) {
      throw new Error(
        "projenTokenSecret is deprecated, please use projenCredentials instead"
      );
    }

    if (options.projenTokenSecret) {
      this.projenCredentials = GithubCredentials.fromPersonalAccessToken({
        secret: options.projenTokenSecret,
      });
    } else if (options.projenCredentials) {
      this.projenCredentials = options.projenCredentials;
    } else {
      this.projenCredentials = GithubCredentials.fromPersonalAccessToken({
        secret: "PROJEN_GITHUB_TOKEN",
      });
    }

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this, options.mergifyOptions);
    }

    if (options.mergeQueue ?? false) {
      this.mergeQueue = new MergeQueue(this, options.mergeQueueOptions);
    }

    if (options.pullRequestLint ?? true) {
      new PullRequestLint(this, options.pullRequestLintOptions);
    }

    if (options.pullRequestBackport ?? false) {
      if (options.mergify) {
        this.project.logger.warn(
          "pullRequestBackport should not be used with mergify as mergify provides its own backport functionality. Please disable pullRequestBackport or mergify."
        );
      }
      new PullRequestBackport(this, options.pullRequestBackportOptions);
    }
  }

  /**
   * All workflows.
   */
  public get workflows(): GithubWorkflow[] {
    const isWorkflow = (c: Component): c is GithubWorkflow =>
      c instanceof GithubWorkflow;
    return this.project.components
      .filter(isWorkflow)
      .sort((w1, w2) => w1.name.localeCompare(w2.name));
  }

  /**
   * Adds a workflow to the project.
   * @param name Name of the workflow
   * @returns a GithubWorkflow instance
   */
  public addWorkflow(name: string) {
    const workflow = new GithubWorkflow(this, name);
    return workflow;
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }

  /**
   * Finds a GitHub workflow by name. Returns `undefined` if the workflow cannot be found.
   * @param name The name of the GitHub workflow
   */
  public tryFindWorkflow(name: string): undefined | GithubWorkflow {
    return this.workflows.find((w) => w.name === name);
  }

  /**
   * Whether downloading from LFS is enabled for this GitHub project
   */
  public get downloadLfs() {
    return this._downloadLfs ?? this.project.gitattributes.hasLfsPatterns;
  }
}
