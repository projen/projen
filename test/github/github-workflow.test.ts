import * as YAML from "yaml";
import { WorkflowSteps } from "../../src/github";
import { GithubWorkflow } from "../../src/github/workflows";
import { synthSnapshot, TestProject } from "../util";

describe("github-workflow", () => {
  const workflowName = "test-workflow";

  test("env is not set by default", () => {
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName);

    const snapshot = synthSnapshot(project);

    const workflow = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );

    expect(workflow.env).toBeUndefined();
  });

  test("env can be set at workflow level", () => {
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName, { env: { FOO: "bar" } });

    const snapshot = synthSnapshot(project);

    const workflow = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );

    console.log(JSON.stringify(workflow, null, 2));

    expect(workflow.env).toEqual({ FOO: "bar" });
  });

  test("concurrency is not set by default", () => {
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName);

    const snapshot = synthSnapshot(project);

    const workflow = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );

    expect(workflow.concurrency).toBeUndefined();
  });

  test("concurrency defaults", () => {
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName, {
      limitConcurrency: true,
    });

    const snapshot = synthSnapshot(project);

    const workflow = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );

    expect(workflow.concurrency).toEqual({
      "cancel-in-progress": false,
      group: "${{ github.workflow }}",
    });
  });

  test("can override concurrency defaults", () => {
    const project = new TestProject();

    new GithubWorkflow(project.github!, workflowName, {
      limitConcurrency: true,
      concurrencyOptions: {
        group: "custom-group",
        cancelInProgress: true,
      },
    });

    const snapshot = synthSnapshot(project);

    const workflow = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );

    expect(workflow.concurrency).toEqual({
      "cancel-in-progress": true,
      group: "custom-group",
    });
  });

  test("can set custom file name", () => {
    const project = new TestProject();

    const fileName = "FoObAr%.yaml";
    new GithubWorkflow(project.github!, workflowName, {
      fileName,
    });

    const snapshot = synthSnapshot(project);
    expect(snapshot[`.github/workflows/${fileName}`]).not.toBeUndefined();
  });

  test("custom file name must use a YAML extension", () => {
    expect(() => {
      const project = new TestProject();

      const fileName = "foobar.txt";
      new GithubWorkflow(project.github!, workflowName, {
        fileName,
      });
    }).toThrowError(
      "GitHub Workflow files must have either a .yml or .yaml file extension, got: foobar.txt"
    );
  });

  test("workflow job calling a reusable workflow", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("call-another-workflow", {
      name: "Some reusable workflow",
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      with: {
        foo: true,
        bar: "example",
      },
      secrets: {
        credentials: "super-secret",
      },
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job calling a reusable workflow & inherited secrets", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("call-another-workflow", {
      name: "Some reusable workflow",
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      secrets: "inherit",
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("working-directory can be set on a Step", () => {
    const project = new TestProject();

    const ghw = new GithubWorkflow(project.github!, workflowName);
    ghw.addJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ run: "ls", workingDirectory: ".github/workflows" }],
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job calling a reusable workflow with a strategy matrix, using domain string value", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("get-matrix", {
      name: "Get a matrix",
      runsOn: ["ubuntu-latest"],
      permissions: {},
      outputs: {
        calculatedFruits: { stepId: "set-fruits", outputName: "fruits" },
      },
      steps: [
        {
          id: "set-fruits",
          name: "Set fruit array output",
          run: `echo "fruits=$(echo '["apple", "orange", "banana"]' | jq -c .)" >> $GITHUB_OUTPUT`,
        },
      ],
    });

    workflow.addJob("use-a-matrix", {
      name: "Some reusable workflow",
      needs: ["get-matrix"],
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      secrets: "inherit",
      strategy: {
        matrix: {
          domain: {
            fruit: "${{ needs.get-matrix.outputs.calculatedFruits }}",
          },
        },
      },
      with: {
        fruit: "${{ matrix.fruit }}",
      },
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job calling a reusable workflow with a strategy matrix", () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const workflow = new GithubWorkflow(project.github!, workflowName);

    workflow.addJob("use-a-matrix", {
      name: "Some reusable workflow",
      uses: "some-user/example-action.yaml@v1",
      permissions: {},
      secrets: "inherit",
      strategy: {
        matrix: {
          domain: {
            fruit: ["apple", "orange", "banana"],
          },
        },
      },
      with: {
        fruit: "${{ matrix.fruit }}",
      },
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
  });

  test("workflow job can be removed after creation", () => {
    const project = new TestProject();

    const workflow = new GithubWorkflow(project.github!, workflowName);
    workflow.addJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ uses: "actions/checkout@v4" }],
    });
    workflow.removeJob("working-dir");

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();

    const workflowYAML = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );
    expect(workflowYAML.jobs["working-dir"]).toBeUndefined();
  });

  test("workflow job can be updated after creation", () => {
    const project = new TestProject();

    const workflow = new GithubWorkflow(project.github!, workflowName);
    workflow.addJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ uses: "actions/checkout@v4" }],
    });
    workflow.updateJob("working-dir", {
      runsOn: ["ubuntu-latest"],
      permissions: {},
      steps: [{ uses: "actions/checkout@v2" }],
    });

    const snapshot = synthSnapshot(project);

    expect(snapshot[`.github/workflows/${workflowName}.yml`]).toMatchSnapshot();
    const workflowYAML = YAML.parse(
      snapshot[`.github/workflows/${workflowName}.yml`]
    );
    expect(workflowYAML.jobs["working-dir"].steps[0].uses).toEqual(
      "actions/checkout@v2"
    );
  });

  describe("reusable actions", () => {
    test("can use an action on a step", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [{ uses: "actions/checkout@v4" }],
      });

      const snapshot = synthSnapshot(project);

      expect(
        snapshot[`.github/workflows/${workflowName}.yml`]
      ).toMatchSnapshot();
    });

    test("can override the action used in a step", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [{ uses: "actions/checkout@v4" }],
      });

      project.github?.actions.set(
        "actions/checkout@v4",
        "replacement/checkout"
      );

      const snapshot = synthSnapshot(project);
      const wf = snapshot[`.github/workflows/${workflowName}.yml`];

      expect(wf).not.toContain("actions/checkout@v4");
      expect(wf).toContain("replacement/checkout");
    });

    test("can override action without explicitly defining version", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [{ uses: "actions/checkout@v4" }],
      });

      project.github?.actions.set("actions/checkout", "replacement/checkout");

      const snapshot = synthSnapshot(project);
      const wf = snapshot[`.github/workflows/${workflowName}.yml`];

      expect(wf).not.toContain("actions/checkout@v4");
      expect(wf).toContain("replacement/checkout");
    });

    test("can override multiple versions and the default", () => {
      const project = new TestProject();

      const ghw = new GithubWorkflow(project.github!, workflowName);
      ghw.addJob("working-dir", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [
          { uses: "actions/checkout@v2" },
          { uses: "actions/checkout@v3" },
          { uses: "actions/checkout@v4" },
        ],
      });

      project.github?.actions.set(
        "actions/checkout@v4",
        "actions/checkout@explicit-v4"
      );
      project.github?.actions.set(
        "actions/checkout",
        "actions/checkout@generic-override"
      );

      const snapshot = synthSnapshot(project);
      const wf = snapshot[`.github/workflows/${workflowName}.yml`];

      expect(wf).toMatchSnapshot();
      expect(wf).not.toContain("actions/checkout@v2");
      expect(wf).not.toContain("actions/checkout@v3");
      expect(wf).not.toContain("actions/checkout@v4");
      expect(wf).toContain("actions/checkout@explicit-v4");
      expect(wf).toContain("actions/checkout@generic-override");
    });
  });

  describe("Download Artifact", () => {
    const downloadWorkflow = "DownloadArtifact";

    test("sets defaults correctly", () => {
      const project = new TestProject();

      const workflow = new GithubWorkflow(project.github!, downloadWorkflow);

      workflow.addJob("downloading-artifact", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [WorkflowSteps.downloadArtifact()],
      });

      const snapshot = synthSnapshot(project)[workflow.file?.path!];

      expect(snapshot).toMatchInlineSnapshot(`
        "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

        name: DownloadArtifact
        on: {}
        jobs:
          downloading-artifact:
            runs-on: ubuntu-latest
            permissions: {}
            steps:
              - name: Download artifact
                uses: actions/download-artifact@v4
        "
      `);
    });

    test("can set values ", () => {
      const project = new TestProject();

      const workflow = new GithubWorkflow(project.github!, downloadWorkflow);

      workflow.addJob("downloading-artifact", {
        runsOn: ["ubuntu-latest"],
        permissions: {},
        steps: [
          WorkflowSteps.downloadArtifact({
            with: {
              name: "foobar",
            },
          }),
        ],
      });

      const snapshot = synthSnapshot(project)[workflow.file?.path!];
      expect(snapshot).toMatchInlineSnapshot(`
        "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

        name: DownloadArtifact
        on: {}
        jobs:
          downloading-artifact:
            runs-on: ubuntu-latest
            permissions: {}
            steps:
              - name: Download artifact
                uses: actions/download-artifact@v4
                with:
                  name: foobar
        "
      `);
    });
  });
});
