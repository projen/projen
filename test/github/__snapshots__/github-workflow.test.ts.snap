// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github-workflow reusable actions can override multiple versions and the default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@generic-override
      - uses: actions/checkout@generic-override
      - uses: actions/checkout@explicit-v4
"
`;

exports[`github-workflow reusable actions can use an action on a step 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@v4
"
`;

exports[`github-workflow workflow job calling a reusable workflow & inherited secrets 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  call-another-workflow:
    name: Some reusable workflow
    permissions: {}
    uses: some-user/example-action.yaml@v1
    secrets: inherit
"
`;

exports[`github-workflow workflow job calling a reusable workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  call-another-workflow:
    name: Some reusable workflow
    permissions: {}
    uses: some-user/example-action.yaml@v1
    with:
      foo: true
      bar: example
    secrets:
      credentials: super-secret
"
`;

exports[`github-workflow workflow job calling a reusable workflow with a strategy matrix 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  use-a-matrix:
    name: Some reusable workflow
    permissions: {}
    uses: some-user/example-action.yaml@v1
    with:
      fruit: \${{ matrix.fruit }}
    secrets: inherit
    strategy:
      matrix:
        fruit:
          - apple
          - orange
          - banana
"
`;

exports[`github-workflow workflow job calling a reusable workflow with a strategy matrix, using domain string value 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  get-matrix:
    name: Get a matrix
    runs-on: ubuntu-latest
    permissions: {}
    outputs:
      calculatedFruits: \${{ steps.set-fruits.outputs.fruits }}
    steps:
      - name: Set fruit array output
        id: set-fruits
        run: echo "fruits=$(echo '["apple", "orange", "banana"]' | jq -c .)" >> $GITHUB_OUTPUT
  use-a-matrix:
    name: Some reusable workflow
    needs: get-matrix
    permissions: {}
    uses: some-user/example-action.yaml@v1
    with:
      fruit: \${{ matrix.fruit }}
    secrets: inherit
    strategy:
      matrix:
        fruit: \${{ needs.get-matrix.outputs.calculatedFruits }}
"
`;

exports[`github-workflow workflow job can be removed after creation 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs: {}
"
`;

exports[`github-workflow workflow job can be updated after creation 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@v2
"
`;

exports[`github-workflow working-directory can be set on a Step 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: test-workflow
on: {}
jobs:
  working-dir:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - run: ls
        working-directory: .github/workflows
"
`;
